### PR TITLE
Snapshot dimension behavior change

### PR DESCRIPTION
**Pull request by b-butler**
`dimensions` can no longer be set, and is dependent on box.Lz. If `box.Lz
== 0` then `dimensions == 2` otherwise it equals 3.

## Description

Changes `ConfigurationData.dimensions` to be unsettable and depend on the snapshot box. If `box.Lz == 0` then `dimensions == 2` otherwise 3.

## Motivation and context

This helps keep behavior consistent with `hoomd.Box` and `hoomd.State.box`.

## How has this been tested?

Has not been tested directly.

## Change log

```
hoomd.ConfigurationData.dimensions cannot be set and depends on box's z dimension.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/master/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

File: CHANGELOG.rst [hunk 1]
```diff
@@ -14,6 +14,9 @@ v3.0.0-beta.2 (not yet released)
 
 *Changed*
 
+- ``hoomd.snapshot.ConfigurationData.dimensions`` is unsettable
+  and depends on the snapshot box. If :math:`box.Lz = 0` then
```
Comment (bdice) [inline CHANGELOG.rst:18]: This is more like code than math.
```suggestion
  and depends on the snapshot box. If ``box.Lz == 0`` then
```

File: hoomd/snapshot.py [hunk 2]
```diff
@@ -31,6 +27,7 @@ def box(self, box):
             raise ValueError(
```
Comment (bdice) [inline hoomd/snapshot.py:27]: On the line above this, I would recommend catching a limited set of exceptions that we know can be thrown by `from_box`.

File: hoomd/snapshot.py [hunk 3]
```diff
@@ -31,6 +27,7 @@ def box(self, box):
...
                 "{} is not convertible to a hoomd.Box object. "
                 "using hoomd.Box.from_box".format(box))
```
Comment (bdice) [inline hoomd/snapshot.py:29]:
```suggestion
                f"{box} is not convertible to a hoomd.Box object "
                f"using hoomd.Box.from_box()")
```

**Review by bdice — CHANGES_REQUESTED** [review]

**Review by joaander — COMMENTED** [review]
Thanks! This looks good to me. Address @bdice's comments and we can merge.

File: hoomd/snapshot.py (hunk 2 above)
Comment (b-butler) [reply hoomd/snapshot.py:27]: There is a
```
except:
     raise
````
in the method like in `freud` so technically we don't know what errors will be raised necessarily. I generally agree with the limited catching of exceptions, but without converting the error in `from_box` to a `ValueError` I don't think we can do this completely safely.

Comment (bdice) [reply hoomd/snapshot.py:27]: Ok, sounds fine.

**Review by bdice — APPROVED** [review]
I pushed a couple small changes to edit the changelog / error message. This is ready to merge.

**Pull request merged**